pkg/ebpf/c/bpf_helpers_custom.h

```diff
@@ -2,13 +2,26 @@
 #define __BPF_HELPERS_CUSTOM__
 
 #include "bpf_cross_compile.h"
+#include "compiler.h"
 
-/* Macro to output debug logs to /sys/kernel/debug/tracing/trace_pipe
+/*
+ * Macro to output debug logs to /sys/kernel/debug/tracing/trace_pipe
+ *
+ * By default it always adds a newline. A ConstantEditor is used to remove
+ * the extra one in linux >= 5.9.
+ *
+ * The trick here is that bpf_trace_printk doesn't seem to care if you pass a size
+ * larger than what you actually print, as long as the format string is null terminated
+ * somewhere and the buffer itself is of the proper size. That way, we can modify the
+ * last character and set it to null.
  */
 #ifdef DEBUG
 #define log_debug(fmt, ...)                                        \
     ({                                                             \
-        char ____fmt[] = fmt;                                      \
+        __u64 ___last_char = '\n';                                 \
+        LOAD_CONSTANT("log_debug_last_character", ___last_char);   \
```

Comment:
I'm worried about the instruction count increase from this change. It every line is a single instruction (and I'm not sure it is), then this goes from 2 to 5 instructions.

Did you explore the idea of patching the instructions on load? That wouldn't require any more instructions.